parsec/core/local_device.py

```diff
@@ -112,51 +153,77 @@ def short_user_display(self) -> str:
     def device_display(self) -> str:
         return self.device_label or str(self.device_id.device_name)
 
-    @property
-    def slug(self) -> str:
-        # Drop the `.keys` suffix
-        return self.key_file_path.stem
-
     @property
     def slughash(self) -> str:
         return sha256(self.slug.encode()).hexdigest()
 
 
-def list_available_devices(config_dir: Path) -> List[AvailableDevice]:
+def _load_legacy_device_file(key_file_path: Path) -> Optional[AvailableDevice]:
+    # For the legacy device files, the slug is contained in the device filename
+    slug = key_file_path.stem
+
     try:
-        candidate_pathes = list(get_devices_dir(config_dir).iterdir())
-    except FileNotFoundError:
-        return []
-
-    # Sanity checks
-    devices = []
-    for device_folder_path in candidate_pathes:
-        slug = device_folder_path.name
-        key_file_path = device_folder_path / f"{slug}.keys"
-
-        try:
-            organization_id, device_id = LocalDevice.load_slug(slug)
-            # Not a valid slug, ignore this folder
-        except ValueError:
-            continue
+        organization_id, device_id = LocalDevice.load_slug(slug)
+    except ValueError:
+        # Not a valid slug, ignore this file
+        return None
+
+    try:
+        data = legacy_key_file_serializer.loads(key_file_path.read_bytes())
+    except (FileNotFoundError, LocalDeviceError):
+        # Not a valid device file, ignore this file
+        return None
+
+    return AvailableDevice(
+        key_file_path=key_file_path,
+        organization_id=organization_id,
+        device_id=device_id,
+        human_handle=data["human_handle"],
+        device_label=data["device_label"],
+        slug=slug,
+    )
 
-        try:
-            data = key_file_serializer.loads(key_file_path.read_bytes())
-        except (FileNotFoundError, LocalDeviceError):
-            # Not a valid device file, ignore this folder
+
+def _load_device_file(key_file_path: Path) -> Optional[AvailableDevice]:
+    try:
+        data = key_file_serializer.loads(key_file_path.read_bytes())
+    except (FileNotFoundError, LocalDeviceError):
+        # Not a valid device file, ignore this folder
+        return None
+
+    return AvailableDevice(
+        key_file_path=key_file_path,
+        organization_id=data["organization_id"],
+        device_id=data["device_id"],
+        human_handle=data["human_handle"],
+        device_label=data["device_label"],
+        slug=data["slug"],
+    )
+
+
+def _iter_available_devices(config_dir: Path) -> Iterator[AvailableDevice]:
+    # Set of seen slugs
+    seen = set()
+    # Loop over `.keys` files in devices directory and subdirectories
+    for key_file_path in get_devices_dir(config_dir).rglob("*.keys"):
```

Comment:
This could be troublesome in case we have multiple files with the same slug (if a user has copied a key file for instance)

We should enforce that the iteration is done in a sorted way so calling multiple time `get_key_file` always provide the same path (and we should document this as well of course ^^)